pkg/controller/certificaterequests/vault/vault_test.go

```diff
@@ -0,0 +1,299 @@
+/*
+Copyright 2019 The Jetstack cert-manager contributors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package vault
+
+import (
+	"context"
+	"errors"
+	"testing"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+
+	"github.com/jetstack/cert-manager/pkg/apis/certmanager"
+	"github.com/jetstack/cert-manager/pkg/apis/certmanager/v1alpha1"
+	testcr "github.com/jetstack/cert-manager/pkg/controller/certificaterequests/test"
+	testpkg "github.com/jetstack/cert-manager/pkg/controller/test"
+	internalvault "github.com/jetstack/cert-manager/pkg/internal/vault"
+	fakevault "github.com/jetstack/cert-manager/pkg/internal/vault/fake"
+	"github.com/jetstack/cert-manager/test/unit/gen"
+)
+
+func TestSign(t *testing.T) {
+	rsaPK := testcr.GenerateRSAPrivateKey(t)
+	caCSR := testcr.GenerateCSR(t, rsaPK)
+
+	testCR := gen.CertificateRequest("test-cr",
+		gen.SetCertificateRequestCSR(caCSR),
+		gen.SetCertificateRequestIsCA(true),
+		gen.SetCertificateRequestDuration(&metav1.Duration{Duration: time.Hour * 24 * 60}),
+		gen.SetCertificateRequestIssuer(v1alpha1.ObjectReference{
+			Name:  "vault-issuer",
+			Group: certmanager.GroupName,
+			Kind:  "Issuer",
+		}),
+	)
+
+	_, rsaPEMCert := testcr.GenerateSelfSignedCertFromCR(t, testCR, rsaPK, time.Hour*24*60)
+
+	tokenSecret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: gen.DefaultTestNamespace,
+			Name:      "token-secret",
+		},
+		Data: map[string][]byte{
+			"my-token-key": []byte("my-secret-token"),
+		},
+	}
+
+	roleSecret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: gen.DefaultTestNamespace,
+			Name:      "role-secret",
+		},
+		Data: map[string][]byte{
+			"my-role-key": []byte("my-secret-role"),
+		},
+	}
+
+	tests := map[string]testT{
+		"a badly formed CSR should report failure": {
+			issuer: gen.Issuer("vault-issuer"),
+			certificateRequest: gen.CertificateRequestFrom(testCR,
+				gen.SetCertificateRequestCSR([]byte("a bad csr")),
+			),
+			builder: &testpkg.Builder{
+				KubeObjects:        []runtime.Object{},
+				CertManagerObjects: []runtime.Object{},
+				ExpectedEvents: []string{
+					`Warning ErrorParsingCSR Failed to decode CSR in spec: error decoding certificate request PEM block`,
+				},
+				CheckFn: testcr.MustNoResponse,
+			},
+			expectedErr: false,
+		},
```

Comment:
I don't see any ExpectedActions here that ensure it was marked as failed?

Reply:
In order to use ExpectedActions, you're going to need to test the actual CertificateRequests controller itself with this issuer loaded. I think we discussed this before, but doing so will save you implementing/having to use additional logic in the CheckFn (and thus make our tests more consistent/easier to write IMO).

Whilst I see your point about testing at a lower level, given the nature of this code relies upon being run as a controller anyway (we absorb errors and return nil due to the controller-like design), given we actually use the same testing strategy as we do when testing controllers (yet can't test the full surface), I feel like so far we've only *lost* granularity in our tests with this.